### ISSUE TITLE
Avoid building transaction in AddChangeIfNeeded of TxBuilder

### ISSUE THREAD

**Issue by visopsys** [echovl/cardano-go#29]
The [AddChangeIfNeeded](https://github.com/echovl/cardano-go/blob/4936c872fbb1f1db4bf04f1242fc180b0fe9843f/tx_builder.go#L77) of TxBuilder calls `tx.build()` inside which does a transaction signing. Calling `tx.Build()` at this stage is an awkward way of verifying transaction, especially when the `build` function mostly does transaction signing. 

Note that after `AddChangeIfNeeded` is called, we still need to build the transaction again and therefore do the signing twice. In general, we should only do transaction signing once when all data is available. 

Can we remove `tx.build()` in `AddChangeIfNeeded` ?

**Comment by visopsys** [comment]
I can see that you need to build all hidden fields (like `AuxiliaryDataHash` or `WitnessSet`) to generate a full transaction to calculate the exact tx bytes used in the fee calculation. 

We can at least avoid double signing by creating placeholder witnesses (note that public key is always 32 bytes and signature is always 64 bytes) in the building process. However, this does not completely solve the problem. 

I am very curious how Cardano's implementation in other languages handle this case.

**Comment by echovl** [comment]
Yes we need the full tx in order to calculate the correct fee,  we can build it with fake data but we need to know how many witnesses are required, this is the approach that cardano-serialization-lib follows.

**Comment by visopsys** [comment]
The problem with current implementation of `AddChangeIfNeeded` is that it assumes all other fields of the tx (e.g. auxiliary data, signers, etc) are present in order to calculate the correct fee. However, I think most users are not aware of this and they might add auxiliary data after calling `AddChangeIfNeeded` and the fee calculation is not correct anymore. 

IMHO, the correct way of doing `AddChangeIfNeeded` is to do it when `Build` is called. The `AddChangeIfNeeded` should only save the recipient address in this function and delay the fee calculation and extra UXTO before the building happens. With this approach, we can handle the case where users add more data after `AddChangeIfNeeded` is caled.

**Comment by echovl** [comment]
You're right that's a better approach, I'm going to work on this. 

That should fix `AddChangeIfNeeded` though we still need something like `buildFakeTx` for `MinFee`.

**Comment by visopsys** [comment]
Yeah, we still need `buildFakeTx`. There seems to be no way around this. This is an awkward way for Cardano to handle this as `MinFee` can, on theory, modify the tx byte size though it does not happen in practice. They should have separated the signature from the transaction body.
